### PR TITLE
Isolating ltac naming out of pretyping + fixing renaming 

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** refactoring, little fixes

We continue the work started by @ppedrot in isolating a module `ExtraEnv` to deal with renaming in `pretyping.ml`. We isolate a new file `globEnv.ml` which includes the `ExtraEnv` API as well as all functions related to a possible interpretation of CIC binding names as Ltac variable names for CIC variable names. So, no more explicit Ltac-specific management in `pretyping.ml`: it is hidden in the new module.

This PR stands for itself but it can a priori be combined with #6090 as well.

We shall later use this infrastructure to develop a fix for tactics-in-term not taking into account renaming done in the context of evars.

In the process, we fix "minor" bugs in interpreting renamings coming from Ltac. An example is:
```
Goal forall a, a = 0.
match goal with
|- (forall x, x = _) => assert (forall n, (fix x n := match n with O => O | S n => x n end) n = n)
end.
```

We also extend the precomputation done in `ExtraEnv` (fa324f9b for #4964) to `new_evar_type`, but:
- using it in `cases.ml` remains to be done;
- the computation of evar contexts can actually be costly especially with the current `KeepUserNameAndRenameExisting` naming strategy (see discussion around #307).

I start to be relatively satisfied with the API but comments are of course welcome.

### CODE REVIEW

File: pretyping/globEnv.mli [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+(** Declare an evar using renaming information *)
+
+val e_new_evar : t -> evar_map ref -> ?src:Evar_kinds.t Loc.located ->
+  ?naming:Misctypes.intro_pattern_naming_expr -> constr -> constr
+
+val e_new_type_evar : t -> evar_map ref -> src:Evar_kinds.t Loc.located -> constr
+
+(** Tell to hide the binding of a variable in the ltac environment;
+    useful e.g. for the dual status of "y" as term and binder in
+    "match y with ..." *)
```
Comment (gares) [inline pretyping/globEnv.mli]: Could you please complete the example? I was not aware of `y` playing a double role here.

Comment (herbelin) [reply pretyping/globEnv.mli]: Done. Tell me if it is clearer.

File: pretyping/globEnv.mli [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+(** Tell to hide the binding of a variable in the ltac environment;
+    useful e.g. for the dual status of [y] as term and binder. This is
+    the case of [match y return p with ... end] which denotes
+    [match z as z return p with ... end] when [y] is bound to a
+    variable [z] and [match t as y return p with ... end] when [y] is
+    bound to a non-variable term [t]. In the latter case, the binding
+    of [y] to [t] should be hidden. *)
+
+val hide_variable : t -> Name.t -> Id.t -> t
```
Comment (gares) [inline pretyping/globEnv.mli]: It is better than before but it is still hard to understand IMO.

In `hide_variable env name id` what is hidden? `name` or `id` or both? And in the example, which subterm (between `p` and `...`) has to be interpreted in the updated env?

File: pretyping/globEnv.mli [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+(** [hide_variable id na] tells to hide the binding of [id] in the
+    ltac environment and to additionally rebind it to [id'] in case [na]
+    is [Name id']. It is useful e.g. for the dual status of [y] as
+    term and binder. This is the case of [match y return p with ... end]
+    which denotes [match z as z return p with ... end] when [y] is
+    bound to a variable [z] and [match t as y return p with ... end]
+    when [y] is bound to a non-variable term [t]. In the latter case,
+    the binding of [y] to [t] should be hidden. *)
+
+val hide_variable : t -> Name.t -> Id.t -> t
```
Comment (herbelin) [inline pretyping/globEnv.mli]: Ah, ok, I then added an explanation of the effect of the function.

Comment (gares) [reply pretyping/globEnv.mli]: I think you got the order of the arguments in `[hide_variable id na]` wrong (wrt the type of the function).

I'd extend the last line as `the binding of [y] to [t] should be hidden in [p].` (assuming I got it right).

**Comment by SkySkimmer** [comment]
A gitlab.yml change somehow made its way into this PR's commit list. Rebase issue?

File: pretyping/globEnv.mli (hunk 3 above)
Comment (herbelin) [reply pretyping/globEnv.mli]: Yep, you get it right (twice). Updated.

**Comment by herbelin** [comment]
> Rebase issue?

Yes, fixed. (Actually the previous rebase was already not exactly the intended one.)

File: pretyping/globEnv.mli [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+open Ltac_pretype
+
+val empty_lvar : ltac_var_map
+
+(** To embed constr in glob_constr *)
+
+val register_constr_interp0 :
+  ('r, 'g, 't) Genarg.genarg_type ->
+    (unbound_ltac_var_map -> env -> evar_map -> types -> 'g -> constr * evar_map) -> unit
+
```
Comment (ppedrot) [inline pretyping/globEnv.mli]: Why did you move the registering function to `GlobEnv`? It belongs semantically to pretyping.

File: pretyping/globEnv.ml [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+      variables pointing to the effective expected name
+    - it incrementally and lazily computes the renaming of rel
+      variables used to build purely-named evar contexts
+*)
+
+type t = {
+  static_env : env;
+  (** For locating indices *)
+  renamed_env : env;
+  (** For name management *)
```
Comment (ppedrot) [inline pretyping/globEnv.ml]: Rather than a full-fledged environment, I think that a rel_context should be enough.

File: pretyping/globEnv.mli [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+val push_rel : evar_map -> rel_declaration -> t -> rel_declaration * t
+val push_rel_context : ?force_names:bool -> evar_map -> rel_context -> t -> rel_context * t
+val push_rec_types : evar_map -> Name.t array * constr array -> t -> Name.t array * t
+
+(** Declare an evar using renaming information *)
+
+val e_new_evar : t -> evar_map ref -> ?src:Evar_kinds.t Loc.located ->
+  ?naming:Misctypes.intro_pattern_naming_expr -> constr -> constr
+
```
Comment (ppedrot) [inline pretyping/globEnv.mli]: I know that the pretyper is still using that style, but I'm not very fond of the imperative API.

File: pretyping/globEnv.ml (hunk 5 above)
Comment (herbelin) [reply pretyping/globEnv.ml]: This is what I had at some time and I changed it for a full env by symmetry. Anyway, in the last version of this PR used in #7300, I only have one `env` and the renaming data is moved to the `extra` flag. I did not update yet this PR because it is always quite tricky to split a PR into subPRs and to maintain them in a consistent state, but I shall do it at some time when the discussion on the contents is complete.

File: pretyping/globEnv.mli (hunk 6 above)
Comment (herbelin) [reply pretyping/globEnv.mli]: Yes, that was by consistency of style. Would you prefer e.g. that I move entirely `pretyping.ml` to a threaded `evd` (or `sigma`) style?

File: pretyping/globEnv.mli (hunk 4 above)
Comment (herbelin) [reply pretyping/globEnv.mli]: I found it nice to have `pretyping` not needing anything specific to `ltac`. So I wanted to move the interpretation function for tactics-in-term (exactly for genargs-in-terms) to `globEnv`. But then, I was needing the `constr_interp` registration known from `GlobEnv`.

At least two possible alternatives:
- to keep ltac in `pretyping` and to provide a `ltac : GlobEnv.t -> ltac_var_map` function for use in `pretyping.ml`
- to move registration before `GlobEnv` in a file which reflects that it is about typing (meaning also moving `ltac_var_map`)

Comment (ppedrot) [reply pretyping/globEnv.mli]: This function is not specific to Ltac, as it is also used by e.g. Ltac2 and (if memory serves) Mtac. The fact that it takes a `GlobEnv` is a deficiency of the API, nothing more, and should be handled using a specific instance of pretyping once we have a generic pretyper.

File: pretyping/globEnv.ml (hunk 5 above)
Comment (ppedrot) [reply pretyping/globEnv.ml]: OK, no problem then.

File: pretyping/globEnv.mli (hunk 6 above)
Comment (ppedrot) [reply pretyping/globEnv.mli]: I'm not pushing strongly for it, that was mostly a remark.

Comment (maximedenes) [reply pretyping/globEnv.mli]: I think it would be good to move to an explicit threading, yes. If not, I would prefer that we don't stay uniform rather than introducing more imperative interfaces.

Comment (ppedrot) [reply pretyping/globEnv.mli]: Note that I'd like that at some point we need to switch to the tactic API.

Comment (herbelin) [reply pretyping/globEnv.mli]: OK, if noone has an opposite preference within the next hours, I shall make `e_new_evar` functional (and maybe with a different names to avoid the collision with the one if `evarutil.ml`).

Comment (ejgallego) [reply pretyping/globEnv.mli]: @herbelin note #6342

**Review by forestjulien — APPROVED** [review]

File: pretyping/globEnv.mli (hunk 4 above)
Comment (ppedrot) [reply pretyping/globEnv.mli]: > My understanding is that, thanks to open recursion, you'd like parts 2 and 3 to be eventually only in plugins (thus getting also rid of registering).

Indeed.

> So, are you saying that you'd prefer parts 2. and 3. of GlobEnv to remain in pretyping.ml (even if it means keeping ltacX-related code in pretyping) and GlobEnv deal only with the renaming for Evars?

Rather, I think that when we have open recursion we merge back the evar-renaming mechanism into pretyping and we move GlobEnv to the Ltac plugin.

> Regarding Ltac2, I was actually surprised that the passing of a renaming substitution through the ist in #7300 did not work out of the box also for Ltac2. Would you like me to investigate this or do you prefer Ltac2 to support a first-order (non-binding) view at variables?

It's a design choice of Ltac2 that term variables are not handled transparently. It has a dubious semantics and a terrible efficiency cost even when not used. I had to hack around the Ltac hardcoded implementation to make it work properly, actually.

**Comment by herbelin** [comment]
I'm answering in the main thread of discussion as it is not strictly about a particular part of the code but about the long-term objectives.

> > So, are you saying that you'd prefer parts 2. and 3. of GlobEnv to remain in pretyping.ml (even if it means keeping ltacX-related code in pretyping) and GlobEnv deal only with the renaming for Evars?
> 
> Rather, I think that when we have open recursion we merge back the evar-renaming mechanism into pretyping and we move GlobEnv to the Ltac plugin.

One of the advantage of having the evar context renaming mechanism outside of `pretyping.ml` is that e.g. `cases.ml` can use it when pushing a variable or generating a new evar. So, it seems to me that we have no choice than to keep it outside of `pretyping.ml`.

Now, what to do with `GlobEnv` as it is in this PR? Would you like I try as soon as now to split it into two parts, say `GlobEnv` for the evar context renaming and, say, `LtacPretyping` for the LtacX-specific extensions (`interp_ltac_variable`, `interp_glob_genarg`, `hide_variable`, which are parts 2 and 3). Or is it ok at the current time to keep it as it is currently?

> It's a design choice of Ltac2 that term variables are not handled transparently. It has a dubious semantics ...

I would precisely find challenging that we could give it a non-dubious semantics. We give a semantics to binders when we write `fun x => x`, why couldn't we give a semantics to `fun x => ltacX:(some-tactic x)` as well (this is what I'm trying to do in #7300).

> ... and a terrible efficiency cost even when not used. I had to hack around the Ltac hardcoded implementation to make it work properly, actually.

Can you give more details?

**Comment by ppedrot** [comment]
> One of the advantage of having the evar context renaming mechanism outside of pretyping.ml is that e.g. cases.ml can use it when pushing a variable or generating a new evar. So, it seems to me that we have no choice than to keep it outside of pretyping.ml.

Yes, but ultimately the pattern-matching compilation should be polymorphic over the type of environments it requires, simply requiring some lightweight API that `GlobEnv` would be an instance of. That the typical horrible OO inheritance style, without the corresponding syntactic sugar.

> Now, what to do with GlobEnv as it is in this PR? Would you like I try as soon as now to split it into two parts, say GlobEnv for the evar context renaming and, say, LtacPretyping for the LtacX-specific extensions (interp_ltac_variable, interp_glob_genarg, hide_variable, which are parts 2 and 3). Or is it ok at the current time to keep it as it is currently?

I'm fine with the current implementation.

> I would precisely find challenging that we could give it a non-dubious semantics. We give a semantics to binders when we write fun x => x, why couldn't we give a semantics to fun x => ltacX:(some-tactic x) as well (this is what I'm trying to do in #7300).

Well, bound variables definitely have a semantics in Ltac2, but not the one of Ltac1. It's just that in `fun x => ltac2:(...)`, the variable `x` is not bound in the **static** environment, i.e. it is not an ML variable. This is required because otherwise, you'd be back to square one with the folkloric bug `clear x; ...`. But it is specified that it is part of the **dynamic** proof environment, so that if you call immediately `Control.hyp @x` at the entrance of the tactic block, then it will be there. Obviously if you mess with the proof state, this might not hold later on, but that's the problem of the tactic writer, not the language fault, and the function call will fail dynamically.

If you wanted to have it statically, you'd definitely need some serious type system à la FreshML, and this is far beyond what LtacX is.

> Can you give more details?

About the performance hit? I know that in Ltac-heavy, @JasonGross style code, renaming term variables on the fly is a major hotspot for no good reason (most of the time). You pay for it each time you go under a binder or perform a term matching, and this ends up representing a huge overhead *even if you never reuse the name for any binder*. AFAICT the feature was introduced by @aspiwack without really thinking about the practical consequences (not even mentioning semantics), and now we're stuck with it, so I'm being cautious.

**Comment by herbelin** [comment]
> Yes, but ultimately the pattern-matching compilation should be polymorphic over the type of environments it requires, simply requiring some lightweight API that GlobEnv would be an instance of. That the typical horrible OO inheritance style, without the corresponding syntactic sugar.

OK for such a long-term approach to `cases.ml`.

> But it is specified that it is part of the dynamic proof environment, so that if you call immediately `Control.hyp @x` at the entrance of the tactic block, then it will be there.

What is the exact claim? Do you mean that it is the responsibility of the user to know what renaming strategy Coq shall use under the hood? For instance, writing `&` for `Control.hyp`, we have here 3 different renaming strategies used by Coq:
```coq
Require Import Ltac2.
Ltac2 f () := constr:((fun x y : nat => ltac2:(intros ?x; exact (&x+&y))) : nat -> nat -> nat -> nat).
Ltac2 g () := constr:((fun x y : nat => ltac2:(refine (constr:(fun x => ltac2:(exact (&x + &y)))))) : nat -> nat -> nat -> nat).
Goal True.
Std.pose None (f ()).
(* n := fun x y _ : nat => x + y *) (* The "KeepExistingNames" strategy *)
Undo.
Std.pose None (g ()).
(* n := ((fun _ y x0 : nat => x0 + y) *) (* The KeepUserNameAndRenameExisting strategy ... *)
Abort.
Section A.
Variable x:nat.
Goal True.
Std.pose None (g ()).
(* n := fun x y : nat => x + y *) (* ... but only if not a section variable *)
```
These renaming strategies are even context-dependent:
```coq
Ltac2 g () := constr:(fun x y : nat => ltac2:(exact (&x+&y))).
Goal True.
Std.pose None (g ()).
(* n := fun x y : nat => x + y *) (* No clash *)
Abort.
Section A.
Variable x:nat.
Goal True.
Std.pose None (g ()).
(* n := fun _ y : nat => x + y *) (* Clash resolved by Evarutil.KeepUserNameAndRenameExistingButSectionNames *)
```
Does the user have to be aware of which strategy is used in every case, and worse, as in the last case, to anticipate the context in which a given defined tactic will be used?

> If you wanted to have it statically, you'd definitely need some serious type system à la FreshML, and this is far beyond what LtacX is.

That's indeed my point. Ltac2 brings types, so the step to add effects to these types is maybe not so much and worth to be investigated, aiming at providing an as-useful-as-possible intuitive structure of bindings. At first view, one needs for instance effects such as "x cleared", "x renamed into y" and "x new_var" with `clear` typed of type `forall x:var -> unit & x cleared`, `intro` of type `forall x:ident -> unit & x new_var` (where the input `x` is the LtacX variable name of a CIC name, the returned `x` is the LtacX variable name of a CIC variable name, and the effect is that LtacX variable `x` changes its type from `ident` to `var` for the rest of the computation). Then, a code such such as `fun x => ltac2:(clear x; exact x)` would be rejected statically.

We may need more kinds of effect, e.g `destruct x` would need an effect of the form `x cleared if a variable` where `x` has here type `constr` and not type `var`, while `intuition` would need an effect `all variables of type var (potentially) made unavailable`, and `match goal` being tricky,  but the idea is to seize the opportunity that Ltac2 is new to think as soon as now at how to statically provide interesting invariants of a tactic programming language that Ltac, by lack of a type system, was unable to provide. After all, if we do not start at some time to explore type systems for binders, even starting in a rudimentary way, how shall we make progresses in the long term. And as far as I can quickly see from this [paper](https://www.cl.cam.ac.uk/~amp12/papers/frepbm/frepbm.pdf), FreshML is not providing the solution for the cases I'm considering here. At first view, it seems more about HOAS than about defining a clear static scoping for the nesting of meta-language binders and of object-language binders, as it is the case for LtacX.

> About the performance hit? I know that in Ltac-heavy, @JasonGross style code, renaming term variables on the fly is a major hotspot for no good reason (most of the time). You pay for it each time you go under a binder or perform a term matching, and this ends up representing a huge overhead *even if you never reuse the name for any binder*. AFAICT the feature was introduced by @aspiwack without really thinking about the practical consequences (not even mentioning semantics), and now we're stuck with it, so I'm being cautious.

Are you talking about the `add_binders` in `Constr_matching` (which I introduced)? It seems to add an overhead of only a log of the number of binders in the term to match, or am I missing something? And, even if costly, an analysis of which variables are really used could be done so as to reduce the cost.

Or are you talking about @aspiwack's `ltac_interp_name` in 2fcc458a, which, similarly, could be optimized so as to collect only those idents which are liable to actually occur in the term to pretype?

**Comment by ppedrot** [comment]
So, everybody agrees with this PR? Nobody against? @herbelin If that's OK, can you rebase it so that I can merge it? Thanks.

**Comment by herbelin** [comment]
> If that's OK, can you rebase it so that I can merge it? Thanks.

Apparently, the failure in fiat-crypto is not spurious (it is about `GlobEnv.hide_variable`). I'll have a look when I find some time.

**Comment by herbelin** [comment]
> Any progress on this PR?

I had fixed the `hide_variable` problem in fiat-crypto but there was a second problem in fiat-crypto which actually revealed a collision between Ltac names and names generated internally in the pattern-matching compilation algorithm. I fixed it and rebased it. I also tried to clarify the role of `ltac_constrs`, `ltac_uconstrs`, `ltac_genargs`, `ltac_idents` (which are used in Ltac2 too as far as I can see) and upgraded `hide_variable` so that it satisfies the corresponding invariant.

The new version seems ok at this time. The next step on my side on the interaction between pretyping and ltacX interpretation, especially regarding binders, shall be #7300.

**Comment by ppedrot** [comment]
@herbelin So this PR is ready, right? It looks like you need to rebase one more time for the CI to go through unluckily, I'll merge it after.

**Comment by ppedrot** [comment]
> I also tried to clarify the role of ltac_constrs, ltac_uconstrs, ltac_genargs, ltac_idents (which are used in Ltac2 too as far as I can see)

FTR, they are not, I specifically had to work around the Ltac-centric implementation to make Ltac2 work correctly.

**Comment by herbelin** [comment]
> It looks like you need to rebase

Done.

> FTR, they are not, I specifically had to work around the Ltac-centric implementation to make Ltac2 work correctly.

At least you use `ltac_genargs`...

By the way, are you sure that you would not eventually prefer to have Ltac2 functions insensitive to the exact names chosen in binders? As far as I could understand, the behavior of a given Ltac2 function might for instance currently differ depending on alpha-conversion of the names used in a goal.

**Comment by ppedrot** [comment]
@herbelin It seems you forgot to push.

> By the way, are you sure that you would not eventually prefer to have Ltac2 functions insensitive to the exact names chosen in binders? As far as I could understand, the behavior of a given Ltac2 function might for instance currently differ depending on alpha-conversion of the names used in a goal.

I am unsure of what you mean by that, but it is indeed a feature that you can observe the bound names of terms. It is required for completeness and people have already asked for this in Ltac1, coming up with ugly workarounds.

**Comment by ppedrot** [comment]
@herbelin ping.

**Comment by herbelin** [comment]
> It seems you forgot to push.

I should definitely take the habit of double-checking.

I pushed-and-checked but CI does not seem more happy, even if apparently with unrelated failures.

> I am unsure of what you mean

I possibly made a confusion with something unrelated to this PR, sorry. I'll clarify my view and come back to you at another occasion.

> but it is indeed a feature that you can observe the bound names of terms. It is required for completeness and people have already asked for this in Ltac1, coming up with ugly workarounds.

In Ltac1, you can observe bound names for a long time already (e.g. in the style `match c with forall n, ?P => idtac n end`). Are we talking about that? If yes, what are you calling "ugly workarounds"?

**Comment by herbelin** [comment]
> I am unsure of what you mean

Actually, what is an example of what I mean:
```coq
Require Import Ltac2.
Ltac2 f y := let z := constr:(fun x : nat => ltac2:(apply $y)) in set $z.
Lemma L x : x = 0.
f &x.
(* n := fun x : nat => x : nat -> nat*)
Abort.
Lemma L x' : x' = 0.
f &x'.
(* n := fun _ : nat => x' : nat -> nat *)
```
The script is not stable by static renaming of `x` into `x'` because of the collision with an internal variable `x` of the `Ltac2` function `f`.

Note that `ltac:(...)` behaves the same:
```coq
Ltac f y := let z := constr:(fun x : nat => ltac:(apply y)) in set z.
Lemma L x : x = 0.
f x.
(* n := fun x : nat => x : nat -> nat *)
```
but it shall behave up to alpha-conversion of internal names with #7300 (I did not check again because a rebase needed but that's the intention).

**Comment by ppedrot** [comment]
@herbelin This is ready to go, AFAIU. The Ltac2 overlay needs to be rebased, but I can probably take care of this directly. The windows failure looks spurious. If you agree, I can proceed to the merge.

**Comment by herbelin** [comment]
@ppedrot: ok to me. Thanks.
